src/libopensc/card-piv.c

```diff
@@ -2992,19 +3032,47 @@ static int piv_match_card(sc_card_t *card)
 	if (card->type == -1)
 		card->type = type;
 
-	card->drv_data = priv;
+	card->drv_data = priv; /* will frre if no match, or pass on to piv_init */
 	priv->aid_file = sc_file_new();
 	priv->selected_obj = -1;
 	priv->pin_preference = 0x80; /* 800-73-3 part 1, table 3 */
 	priv->logged_in = SC_PIN_STATE_UNKNOWN;
 	priv->tries_left = 10; /* will assume OK at start */
+	priv->pstate = PIV_STATE_MATCH;
 
 	/* Some objects will only be present if Histroy object says so */
 	for (i=0; i < PIV_OBJ_LAST_ENUM -1; i++)
 		if(piv_objects[i].flags & PIV_OBJECT_NOT_PRESENT)
 			priv->obj_cache[i].flags |= PIV_OBJ_CACHE_NOT_PRESENT;
 
+	sc_lock(card);
```

Comment:
You should not assume that `init()` is always followed by `match_card()`, please remove this lock. In the worst case you need to re-do in `init()` what you've already sent in `match_card()`.

Reply:
I think it is the other way around. A successful `piv_match_card` is always followed by `piv_init` 
It also assumes `card->dvr_data` allocated in a successful `piv_match_card` is passed to `piv_init`

By setting `card->dvr_data` in piv_match allows any piv routines t have access to a valid ``card->dvr_data` 

The lock in line 3048 in `piv_init`  is unlocked in:
```
3204         sc_unlock(card) ; /* obtained in piv_match */
```

Reply:
Yes, please don't do it. If you want a valid `card->drv_data` while executing `piv_match_card`, then `sc_lock` **and** `sc_unlock` in `piv_match_card`.

Reply:
The point was to pass from match to init  both the card->drv_data and the lock so any commands done during match don't have to be reissued again, and there can be no interference from other processes between match and init. 

We really need to look at how match is done for applications. PIV and OpenPGP are really applications drivers and we need to have a way to support them better. See private note from Martin I forwarded to you about.     https://github.com/makinako/OpenFIPS201

Reply:
I know that this locking avoids issuing an additional SELECT. [At some point](https://github.com/OpenSC/OpenSC/commit/74ec7b04ffda539819eb8bfd6a0e766e69d8c946#diff-d41bbe692d0f897375a44b6b3929d83aR174) I've done this, too. However, `match_card` was designed to not allocate or lock any internal resources. Other Applet-based card drivers have the same problem and they simply issue an additional SELECT. You can avoid the SELECT in `match_card` if you maintain a list of ATRs of cards that are known to have a PIV applet.

Also, locking is typically done in the PKCS#15 or the PKCS#11 layer, since you have more knowledge about the context that defines what is a transaction. In my opinion the locking should happen where unlocking is done, too. Otherwise we will loose track at some point and lock the token forever. So please do this only when you're forced to.

Reply:
Comment on a few comments in this thread - It is Dangerous to base any PIV (Or other) "card application" identification on ATR alone (or at all), since most cards now support multiple application "Applets" and come with scores of applications in "ready to install" ROM.  PIV requires only one (registered) RID/AID and supports a discovery object, other schemes support other registered/standardised methods.  People need to use the method set out in the standards/doco rather than trying to second guess by the (albeit easier)  ATR. In fact ATR should only be a last worst case option, and it only workled in the past because most cards were single application (which has now changed).

Reply:
I can't find any statement that says you *should* use an ATR. But, as a matter of fact, some cards come with a specific ATR, so one may use it. As example look at [SC-HSM](https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/card-sc-hsm.c#L228-L250)'s implementation:
```
static int sc_hsm_match_card(struct sc_card *card)
{
	sc_path_t path;
	int i, r, type = 0;

	i = _sc_match_atr(card, sc_hsm_atrs, &type);
	if (i >= 0 && type != SC_CARD_TYPE_SC_HSM_SOC) {
		card->type = type;
		return 1;
	}

	sc_path_set(&path, SC_PATH_TYPE_DF_NAME, sc_hsm_aid.value, sc_hsm_aid.len, 0, 0);
	r = sc_hsm_select_file(card, &path, NULL);
	LOG_TEST_RET(card->ctx, r, "Could not select SmartCard-HSM application");

	if (type == SC_CARD_TYPE_SC_HSM_SOC) {
		card->type = SC_CARD_TYPE_SC_HSM_SOC;
	} else {
		card->type = SC_CARD_TYPE_SC_HSM;
	}

	return 1;
}
```
It's fast in some cases (matching an ATR), and it's reliable in all of the cases (sometimes needing a SELECT).